File: py/890.py
import unittest


class Solution:
    def isMatch(self, word, pattern):
        d1 = {}
        d2 = {}
        for i in range(0, len(word)):
            c1 = word[i]
            c2 = pattern[i]
            if d1.get(c1) != None:
                if d1.get(c1) != c2:
                    return False
            elif d2.get(c2) != None:
                return False
            else:
                d1[c1] = c2
                d2[c2] = c1
        return True

    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        return list(filter(lambda word: self.isMatch(word, pattern), words))


class SolutionUnittest(unittest.TestCase):
    def test_findAndReplacePattern(self):
        s = Solution()
        words = ["abc", "deq", "mee", "aqq", "dkd", "ccc"]
        pattern = "abb"
        result = ["mee", "aqq"]
        self.assertListEqual(s.findAndReplacePattern(words, pattern), result)

if __name__ == "__main__":
    unittest.main()

File: js/890.js
var isMatch = function (word, pattern) {
  if (word.length !== pattern.length) {
    return false
  }

  const map = new Map
  const map2 = new Map
  word = Array.from(word)
  for (let i = 0; i < word.length; i++) {
    const c1 = word[i]
    const c2 = pattern[i]
    if (map.has(c1)) {
      if (map.get(c1) !== c2) {
        return false
      }
    } else if (map2.has(c2)) {
      return false
    } else {
      map.set(c1, c2)
      map2.set(c2, c1)
    }
  }
  return true
}

/**
 * @param {string[]} words
 * @param {string} pattern
 * @return {string[]}
 */
var findAndReplacePattern = function (words, pattern) {

  const result = []
  for (let i = 0; i < words.length; i++) {
    if (isMatch(words[i], pattern)) {
      result.push(words[i])
    }
  }
  return result
};

var assert = require('assert');
var words = ["abc", "deq", "mee", "aqq", "dkd", "ccc"],
  pattern = "abb"
var result = ["mee", "aqq"]
assert.deepEqual(findAndReplacePattern(words, pattern), result)